### PR TITLE
inference engine related design

### LINKED ISSUE #10028: any engine for inference subgraph acceleration naive design
# architecture

- frontend to mark the subgraphs that should be optimized by x engine
- inference preparation to get the subgraph, change program desc by replacing the subgraph with an x engine op
- engine op to build an x engine and execute it like a normal operator.

# phrases

## frontend
- manually partition graph
- TODO LATTER automatically partition graph

Some initital ideas, just add some special `with-block` just for inference

```python
a = op0(b, c)

with infer.accelerate_by_tensorrt:
    a1 = op1(b, c)
    a2 = op1(b, c)

c = op2(b, c)

with infer.accelerate_by_some_other_engine:
    a3 = op3(c)
    ...
```

## backend
- inference prepare
  - partition graph and transform the infer program desc
  - tensorrt_engine_op.build
     - convert: input subgraph's block desc, add TensorRT layer into TensorRT engine
         - transform weight format from fluid to tensorrt 
         - add tensorrt layer
- inference execute
  - for op in the infer program desc
     - op.run
 
# x engine
- get a subgraph's block desc (from attribute)
- build x engine once
- execute x engine any times

# x op
- construct x op from x network

# convert
construct x network from a subgraph's block desc

### PR DESCRIPTION
**Pull request by Superjomn**
fixes: https://github.com/PaddlePaddle/Paddle/issues/10028

### CODE REVIEW

**Comment by panyx0718** [comment]
Have we verified the performance of using tensorrt as a sub-graph?

**Comment by Superjomn** [comment]
We will get a benchmark next week.

File: doc/fluid/design/inference_engine.md [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+
+## Engine-related Design
+
+###  EngineOp                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        
+
+`EngineOp` is just a normal Fluid operator, which has an attribute called `subblock` to get the Fluid description about a sub-block.
```
Comment (luotao1) [inline doc/fluid/design/inference_engine.md]: subblock->sub_block

File: doc/fluid/design/inference_engine.md [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+    }
+    engine.Execute();
+    
+    const auto& output0 = engine.buffer("output0");
+    // ...
+}
+*/
+enum class DeviceType {
+    CPU = 0,
+    GPU
```
Comment (luotao1) [inline doc/fluid/design/inference_engine.md]: GPU=1?

File: doc/fluid/design/inference_engine.md [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+
+#define REGISTER_INPUT_CONVERTER(in_op_type__, out_op_type__, Conveter__) \
+    EngineInputConveterBase::Register<Conveter__>(#in_op_type__ "_to_" #out_op_type__);
+
+REGISTER_INPUT_CONVERTER(RNNOp, xEngineOp, RNN2xEngineConveter);
+```
+
+The `EngineOutputConvertOp` is similar.
+
+### Optimizer for sub-block
```
Comment (luotao1) [inline doc/fluid/design/inference_engine.md]: Optimizer->Transpiler

File: doc/fluid/design/inference_engine.md [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+
+```c++
+// The InferenceOptimizers input a program desc and output a block desc.
```
Comment (luotao1) [inline doc/fluid/design/inference_engine.md]: input a program desc, but output maybe a series of sub-block desc

File: doc/fluid/design/inference_engine.md [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+// Different implementations will rewrite the original program desc by different logics.
+// There might be many different optimizers, such as
+// - CleanUselessOptimizer
+// - PruneOpOptimizer
```
Comment (luotao1) [inline doc/fluid/design/inference_engine.md]: what are CleanUselessOptimizer and PruneOpOptimizer ?
We already have `prune` method of inference. see `[paddle\fluid\framework\prune.cc](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/framework/prune.cc)`

**Comment by luotao1** [comment]
Are all of above implemented and run in C++ end?
Following are current C++ inference logic. 
```
inference_program = paddle::inference::Load(&executor, scope, dirname);
executor.Run(*inference_program, scope, ...)
```
Thus, how to use `inference engine`?
```
inference_program = paddle::inference::Load(&executor, scope, dirname);
inference_engine_program = paddle::inference::transplier(inference_program, engine="tensorrt");
executor.Run(*inference_engine_program, scope, ...)
```

File: doc/fluid/design/inference_engine.md [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+## Partition the Sub-blocks supported by a Specific Engine
+
+As mentioned above, one engine can support a partition of Fluid operators, the sub-block dispatched should be composed by the operators this engine fully supports.
+
+The Inference framework needs a mechanism to mark the sub-block and deliver it to an engine.
+
+We use a `with-statement` to mark the sub-block as follows.
+
+```python
+with infer.power_by_engine('tensorrt'):
```
[the remark this answers is not in the document]
Comment (Superjomn) [reply doc/fluid/design/inference_engine.md]: `infer` is a module.

`import paddle.inference as infer`

File: doc/fluid/design/inference_engine.md [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+    o = some_op()
+    o = some_op()
```
[the remark this answers is not in the document]
Comment (Superjomn) [reply doc/fluid/design/inference_engine.md]: No practical meaning, just shows that there are several operators there.

File: doc/fluid/design/inference_engine.md (hunk 2 above)
Comment (Superjomn) [reply doc/fluid/design/inference_engine.md]: The enum syntax just needs to set the first element, and following elements will increase automatically.

File: doc/fluid/design/inference_engine.md (hunk 3 above)
Comment (Superjomn) [reply doc/fluid/design/inference_engine.md]: An optimizer is not a Transpiler. It corresponds to the optimization in a compiler.

File: doc/fluid/design/inference_engine.md (hunk 4 above)
Comment (Superjomn) [reply doc/fluid/design/inference_engine.md]: Input a program desc, output a program desc with several newly inserted `EngineOp` with their attribute set with the sub-blocks.

File: doc/fluid/design/inference_engine.md (hunk 5 above)
Comment (Superjomn) [reply doc/fluid/design/inference_engine.md]: Yes, I this a factory pattern of Operators is a better interface, maybe we'd better refactor those codes.

**Comment by Superjomn** [comment]
The inference might have its own `executor` implementation, so there might be some more consideration about the SDK.

The Anakin and MDL team will join together to design the inference SDK, and there might be some futher designs about these issues. @luotao1

File: doc/fluid/design/inference_engine.md [hunk 8]
```diff
@@ -0,0 +1,254 @@
+# Utilize Engines to Accelerate Inference
```
Comment (wangkuiyi) [inline doc/fluid/design/inference_engine.md]: 这里的engines指的是什么呢？

File: doc/fluid/design/inference_engine.md [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+
+The inference phase need to support some special hardware for acceleration,
```
Comment (wangkuiyi) [inline doc/fluid/design/inference_engine.md]: > The inference phase need to support some special hardware

=>

> We want to utilize DL chips to accelerate the inference of Fluid models.

File: doc/fluid/design/inference_engine.md (hunk 8 above)
Comment (wangkuiyi) [reply doc/fluid/design/inference_engine.md]: 看上去是指 TensorRT？我看到后面提出一个base class，也在另一个code的PR里看到了这个base class。这是为了将来derive除了 TensorRT 之外的其他的“engine”对应的class吗？

**Review by wangkuiyi — CHANGES_REQUESTED** [review]
不好意思，从这个design pr和另一个code pr里，我都没有能领会这个设计的意图。视频会议一下吧。

File: doc/fluid/design/inference_engine.md (hunk 8 above)
Comment (Superjomn) [reply doc/fluid/design/inference_engine.md]: TensorRT, Anajin, 或者其他类似自带完整优化的库

**Comment by luotao1** [comment]
感谢您给PaddlePaddle贡献文档。由于文档已迁移至[FluidDoc](https://github.com/PaddlePaddle/FluidDoc) repo，因此关闭您的PR，欢迎您向[FluidDoc](https://github.com/PaddlePaddle/FluidDoc) Repo贡献文档。
Thanks for contributing to PaddlePaddle! Since documents have been moved to [FluidDoc](https://github.com/PaddlePaddle/FluidDoc) repo, we close this PR. Welcome to contribute to [FluidDoc](https://github.com/PaddlePaddle/FluidDoc) repo.

**Pull request closed without merging**